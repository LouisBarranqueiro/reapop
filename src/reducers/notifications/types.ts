import {POSITIONS, STATUSES} from '../../constants'

export interface NotificationButton {
    name: string
    primary?: boolean
    onClick?: (...args: any[]) => void
}

export type Status = typeof STATUSES[keyof typeof STATUSES]
export type Position = typeof POSITIONS[keyof typeof POSITIONS]

export interface Notification {
    id: string
    title?: string
    message?: string
    status: Status
    position: Position
    buttons: NotificationButton[]
    image?: string
    dismissAfter?: number
    dismissible?: boolean
    onAdd?: (...args: any[]) => void
    onDismiss?: (...args: any[]) => void
    showDismissButton?: boolean
    allowHTML?: boolean
    [index: string]: any
}

export type NewNotification = Partial<Notification>
