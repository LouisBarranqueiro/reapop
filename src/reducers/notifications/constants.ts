export enum NotificationAction {
    UpsertNotification = 'reapop/upsertNotification',
    DismissNotification = 'reapop/dismissNotification',
    DismissNotifications = 'reapop/dismissNotifications',
}
