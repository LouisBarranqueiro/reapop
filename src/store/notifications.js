import {treatNotification, preloadImage} from '../helpers';
import {DEFAULT_NOTIFICATION} from '../constants';

// An array to store notifications object
const INITIAL_STATE = [];
// Action types
const ADD_NOTIFICATION = 'ADD_NOTIFICATION';
const UPDATE_NOTIFICATION = 'UPDATE_NOTIFICATION';
const REMOVE_NOTIFICATION = 'REMOVE_NOTIFICATION';
const REMOVE_NOTIFICATIONS = 'REMOVE_NOTIFICATIONS';

/**
 * Add a notification (thunk action creator)
 *
 * We use a thunk here to create an ADD_NOTIFICATION action
 * and only return the notification object.
 * @param {Object} notification
 * @returns {Object} notification
 */
export const addNotification = (notification) => (dispatch) => {
  if (!notification.id) {
    notification.id = new Date().getTime();
  }
  notification = treatNotification(notification);
  // if there is an image, we preload it
  // and add notification when image is loaded
  if (notification.image) {
    preloadImage(notification.image, dispatch.bind(this, _addNotification(notification)));
  }
  else {
    dispatch(_addNotification(notification));
  }
  return notification;
};

/**
 * Add a notification (action creator)
 *
 * @param {Object} notification
 * @returns {{type: string, payload: {Object}}}
 * @private
 */
function _addNotification(notification) {
  return {
    type: ADD_NOTIFICATION,
    payload: notification
  };
}

/**
 * Update a notification (thunk action creator)
 *
 * We use a thunk here to create an UPDATE_NOTIFICATION action
 * and only return the notification object.
 * @param {Object} notification
 * @returns {Object} notification
 */
export const updateNotification = (notification) => (dispatch, getState) => {
  if (!notification.id) {
    throw new Error('A notification must have an `id` property to be updated');
  }

  const notifications = getState().notifications;
  const index = notifications.findIndex((oldNotification) => oldNotification.id === notification.id);
  const currNotification = notifications[index];

  notification = treatNotification(notification);

  // if image is different, then we preload it
  // and update notification when image is loaded
  if (notification.image && (!currNotification.image || (currNotification.image &&
    notification.image !== currNotification.image))) {
    preloadImage(notification.image, dispatch.bind(this, _updateNotification(notification)));
  }
  else {
    dispatch(_updateNotification(notification));
  }
  return notification;
};

/**
 * Update a notification (action creator)
 *
 * @param {Object} notification
 * @returns {{type: string, payload: {Object}}}
 * @private
 */
function _updateNotification(notification) {
  return {
    type: UPDATE_NOTIFICATION,
    payload: notification
  };
}

/**
 * Remove a notification (action creator)
 *
 * @param {Object} notification
 * @returns {{type: string, payload: {Object}}}
 */
export function removeNotification(notification) {
  return {
    type: REMOVE_NOTIFICATION,
    payload: notification
  };
}

/**
 * Remove all notifications (action creator)
 *
 * @returns {{type: string}}
 */
export function removeNotifications() {
  return {
    type: REMOVE_NOTIFICATIONS
  };
}

// Action creators
export const actions = {
  addNotification: addNotification,
  updateNotification: updateNotification,
  removeNotification: removeNotification,
  removeNotifications: removeNotifications
};

// Actions types
export const types = {
  ADD_NOTIFICATION: ADD_NOTIFICATION,
  UPDATE_NOTIFICATION: UPDATE_NOTIFICATION,
  REMOVE_NOTIFICATION: REMOVE_NOTIFICATION,
  REMOVE_NOTIFICATIONS: REMOVE_NOTIFICATIONS
};

// Reducers
export default (defaultNotification = DEFAULT_NOTIFICATION) => {
  return (state = INITIAL_STATE, {type, payload}) => {
    switch (type) {
      case ADD_NOTIFICATION:
        const notification = Object.assign({}, defaultNotification, payload);
        return [...state, notification];
      case UPDATE_NOTIFICATION:
        // get index of the notification
        const index = state.findIndex((notification) => notification.id === payload.id);
        // replace the old notification by the new one
        state[index] = Object.assign({}, defaultNotification, payload);
        return [...state];
      case REMOVE_NOTIFICATION:
        return state.filter((notification) => notification.id !== payload);
      case REMOVE_NOTIFICATIONS:
        return [];
      default:
        return state;
    }
  };
};
