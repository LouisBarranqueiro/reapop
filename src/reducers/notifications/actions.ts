import {NotificationAction} from './constants'
import {NewNotification, Notification, Status} from './types'
import {isObject} from '../../utils'
import {prepareNotification} from '../../services/notifications'

export type UpsertNotificationAction = {
    type: NotificationAction.UpsertNotification
    payload: Notification
}

export type DismissNotificationAction = {
    type: NotificationAction.DismissNotification
    payload: string
}

export type DismissNotificationsAction = {
    type: NotificationAction.DismissNotifications
}

function upsertNotification(notification: Partial<Notification>): UpsertNotificationAction
function upsertNotification(message: string, options?: Partial<Notification>): UpsertNotificationAction
function upsertNotification(message: string, status: Status, options?: Partial<Notification>): UpsertNotificationAction
function upsertNotification(...args: unknown[]): UpsertNotificationAction {
    const lastArg = args[args.length - 1]
    let notification: NewNotification = {}

    if (typeof args[0] === 'string') {
        notification.message = args[0] as string
    }
    if (typeof args[1] === 'string') {
        notification.status = args[1] as Status
    }
    if (isObject(lastArg)) {
        notification = {...notification, ...(lastArg as NewNotification)}
    }

    return {
        type: NotificationAction.UpsertNotification,
        payload: prepareNotification(notification as NewNotification | Notification),
    }
}

export const dismissNotification = (notificationId: string): DismissNotificationAction => ({
    type: NotificationAction.DismissNotification,
    payload: notificationId,
})

export const dismissNotifications = (): DismissNotificationsAction => ({
    type: NotificationAction.DismissNotifications,
})

export const notify = upsertNotification

export type NotificationActions = UpsertNotificationAction | DismissNotificationAction | DismissNotificationsAction
