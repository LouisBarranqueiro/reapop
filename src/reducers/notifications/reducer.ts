import {NotificationAction} from './constants'
import {DismissNotificationAction, NotificationActions, UpsertNotificationAction} from './actions'
import {Notification} from './types'

type InitReduxAction = {
    type: string
}

const INITIAL_STATE: Notification[] = []

const notificationsReducer = () => {
    return (state = INITIAL_STATE, action: NotificationActions | InitReduxAction): Notification[] => {
        switch (action.type) {
            case NotificationAction.UpsertNotification: {
                const payload = (action as UpsertNotificationAction).payload
                if (~state.findIndex((notif) => notif.id === payload.id)) {
                    return state.map((notif) => (notif.id === payload.id ? payload : notif))
                }
                return [...state, payload]
            }
            case NotificationAction.DismissNotification: {
                const payload = (action as DismissNotificationAction).payload
                return state.filter((notif) => notif.id !== payload)
            }
            case NotificationAction.DismissNotifications:
                return []
            default:
                return state
        }
    }
}

export default notificationsReducer
