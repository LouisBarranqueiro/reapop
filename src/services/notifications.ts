import {CONFIG, NotificationConfig, STATUSES} from '../constants'
import {NewNotification, Notification} from '../reducers/notifications/types'
import {clone} from '../utils'

export const setUpNotifications = (props: Partial<NotificationConfig>) => {
    for (const key in props) {
        const propName = key as keyof NotificationConfig
        CONFIG[propName] = props[propName] as any
    }
}

export const resetNotificationsConfig = () => {
    CONFIG.defaultProps = {}
    CONFIG.customizeNotification = null
    CONFIG.generateId = null
}

export const generateId = () => Math.random().toString(36).slice(2, 7)

export const prepareNotification = (notification: NewNotification): Notification => {
    const notif = clone(notification)
    const defaultProps: Partial<Notification> = {
        ...CONFIG.defaultProps,
        status: STATUSES.none,
        buttons: [],
    }

    if (notif.id === undefined) {
        notif.id = CONFIG.generateId ? CONFIG.generateId(notification) : generateId()
    }

    Object.entries(defaultProps).map(([prop, defaultValue]) => {
        if (notif[prop] === undefined) {
            notif[prop] = defaultValue
        }
    })

    if (typeof CONFIG.customizeNotification === 'function') {
        CONFIG.customizeNotification(notif)
    }
    return notif as Notification
}
