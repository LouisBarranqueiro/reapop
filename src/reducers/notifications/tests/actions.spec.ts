import {dismissNotification, dismissNotifications, notify} from '../actions'
import {setUpNotifications} from '../../..'
import {resetNotificationsConfig} from '../../../services/notifications'

describe('notify()', () => {
    it.each([
        [[{message: 'nice message', allowHTML: true}]],
        [['nice message']],
        [['nice message', {allowHTML: true}]],
        [['nice message', 'error']],
        [['nice message', 'error', {allowHTML: true}]],
    ])('should return action to upsert notification (%s)', (args) => {
        setUpNotifications({
            generateId: () => '1',
        })
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        expect(notify(...args)).toMatchSnapshot()
        resetNotificationsConfig()
    })
})

describe('dismissNotification()', () => {
    it('should return action to dismiss notification', () => {
        expect(dismissNotification('1')).toMatchSnapshot()
    })
})

describe('dismissNotifications()', () => {
    it('should return action to dismiss notifications', () => {
        expect(dismissNotifications()).toMatchSnapshot()
    })
})
