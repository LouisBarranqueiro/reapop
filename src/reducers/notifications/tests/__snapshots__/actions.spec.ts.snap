// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dismissNotification() should return action to dismiss notification 1`] = `
Object {
  "payload": "1",
  "type": "reapop/dismissNotification",
}
`;

exports[`dismissNotifications() should return action to dismiss notifications 1`] = `
Object {
  "type": "reapop/dismissNotifications",
}
`;

exports[`notify() should return action to upsert notification ([ 'nice message' ]) 1`] = `
Object {
  "payload": Object {
    "buttons": Array [],
    "id": "1",
    "message": "nice message",
    "status": "none",
  },
  "type": "reapop/upsertNotification",
}
`;

exports[`notify() should return action to upsert notification ([ 'nice message', 'error' ]) 1`] = `
Object {
  "payload": Object {
    "buttons": Array [],
    "id": "1",
    "message": "nice message",
    "status": "error",
  },
  "type": "reapop/upsertNotification",
}
`;

exports[`notify() should return action to upsert notification ([ 'nice message', 'error', [Object] ]) 1`] = `
Object {
  "payload": Object {
    "allowHTML": true,
    "buttons": Array [],
    "id": "1",
    "message": "nice message",
    "status": "error",
  },
  "type": "reapop/upsertNotification",
}
`;

exports[`notify() should return action to upsert notification ([ 'nice message', [Object] ]) 1`] = `
Object {
  "payload": Object {
    "allowHTML": true,
    "buttons": Array [],
    "id": "1",
    "message": "nice message",
    "status": "none",
  },
  "type": "reapop/upsertNotification",
}
`;

exports[`notify() should return action to upsert notification ([ [Object] ]) 1`] = `
Object {
  "payload": Object {
    "allowHTML": true,
    "buttons": Array [],
    "id": "1",
    "message": "nice message",
    "status": "none",
  },
  "type": "reapop/upsertNotification",
}
`;
